### PR TITLE
[3/3] audioprotopnet: ConvNeXt + AudioProtoPNet + AudioProtoPNetSED models

### PR DESCRIPTION
**Pull request by nkundiushuti**
Add ConvNeXt, AudioProtoPNet, AudioProtoPNet SED, prototype-based 
  classification, and two-stage training

  New models

  ConvNeXt (avex/models/convnext.py)
  - Plain ConvNextForImageClassification audio classifier (1-channel mel
  spectrogram input)
  - Architecture and mel parameters loaded from packaged convnext_base.yml via
  ConvNextCfg — no hardcoded constants
  - backbone property exposes model.convnext (encoder only) for correct
  freeze/unfreeze during two-stage training
  - Hook-based embedding extraction from the 4 encoder stages, compatible with
  all existing probes
  - Transparent PyTorch-Lightning checkpoint loading (strips model.model. /
  model._orig_mod.model. prefixes)
  
  AudioProtoPNet (avex/models/audioprotopnet.py)
  - ConvNeXt backbone + prototype classification head, loaded from HuggingFace
  (DBD-research-group/AudioProtoPNet-{1,5,10,20}-BirdSet-XCL) 
  - backbone property exposes model.model.backbone for correct freeze targeting
  - Exposes ebird_codes (raw eBird codes) and label_mapping (common names via
  bundled eBird taxonomy)
  - Calls ModelBase.__init__ directly to skip the ConvNeXt mel pipeline

  AudioProtoPNet SED (avex/models/audioprotopnet_sed.py)
  - ConvNeXt-Base backbone + 2× bilinear add-on layers + spatial cosine
  prototype head for sound event detection
  - forward() — clip-level logits [B, C] via global H×W max-pool
  - forward_frames() — per-frame sigmoid probabilities [B, T, C] via frequency
  max-pool only; temporal resolution ≈7.6 Hz at 32 kHz
  - Checkpoint loaded from
  gs://representation-learning/models/sed/audioprotopnet-20/ via
  universal_torch_load (cached locally; same pattern as BEATs)
  - _LinearLayerWithoutNegativeConnections: block-diagonal per-class
  non-negative linear layer; weights initialised with kaiming_uniform_
  - Mel preprocessing via ConvNextCfg; backbone architecture from packaged
  convnext_base.yml — no HuggingFace network calls beyond the checkpoint itself

  Config

  ConvNextCfg (avex/configs.py)
  - New Pydantic model for mel preprocessing parameters (sample_rate, n_fft,
  hop_length, n_mels, norm_mean, norm_std); defaults match BirdSet XCL training
  setup
  - Consistent with BEATsConfig pattern; packaged YAMLs carry a top-level
  convnext_cfg block
  - convnext_cfg and checkpoint_dir added to ModelSpec and
  _add_model_spec_params so user YAMLs flow through the factory correctly

  eBird taxonomy (avex/data/)

  - Bundled ebird_taxonomy.json — 18,130 taxa (eBird v2021 + v2025), superset of
   all 9,736 BirdSet XCL species
  - load() with @lru_cache for zero-cost repeated access

  New losses (avex/training/losses.py)

  - AsymmetricLoss (Ridnik et al., ICCV 2021) — separate focusing parameters for
   positives/negatives, probability margin clipping; registered as "asymmetric"
  / "asl" in build_criterion
  - OrthogonalLoss — encourages prototype vectors within the same class to be
  orthogonal (‖P_c Pᵀ_c − I‖_F / numel)

  Prototypical probe (avex/models/probes/prototypical_probe.py)

  - New "prototypical" probe type alongside linear, mlp, attention, lstm,
  transformer
  - Architecture: cosine-similarity prototype layer → [0, 1] shift →
  non-negative linear classifier
  - num_prototypes_per_class added to ProbeConfig

  - New "prototypical" probe type alongside linear, mlp, attention, lstm, transformer
  - Architecture: cosine-similarity prototype layer → [0, 1] shift → non-negative linear classifier
  - num_prototypes_per_class added to ProbeConfig

  Example

  examples/audioprotopnet_species_detection.py — clip-level and frame-level species detection on a directory of audio
  files; --demo mode generates synthetic audio with no real recordings needed

  Tests

  - 24 unit tests for ConvNeXt and AudioProtoPNet (architecture, label mapping, embedding extraction, hook
  registration, layer discovery)
  - 29 unit tests for AudioProtoPNet SED (forward shapes, frame-level properties, batch independence, cosine
  activation, _LinearLayerWithoutNegativeConnections, registry)

### CODE REVIEW

File: avex/data/ebird_taxonomy.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+"""eBird species taxonomy bundled with the package.
+
+Provides a mapping from eBird species codes (e.g. ``"ostric2"``) to common
+and scientific names.  The JSON file covers the full eBird taxonomy
+(v2021 + v2025) — 18 130 taxa — which is a superset of the 9 736 species used
```
Comment (benjaminsshoffman) [inline avex/data/ebird_taxonomy.py]: These taxonomies definitely shouldn't be combined. Each year they revise the full ebird/clements taxonomy and so there will be contradictions between the two. The right way to do this is to look, for each model, at which taxonomy they used, and map with the correct taxonomy version.

File: examples/audioprotopnet_species_detection.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
+"""
+AudioProtoPNet species detection example.
+
+Demonstrates two complementary detection modes on a directory of audio files:
+
+  Clip-level (AudioProtoPNet)
```
Comment (benjaminsshoffman) [inline examples/audioprotopnet_species_detection.py]: The model we trained for the SED project also makes clip-level predictions, however the pooling there is not max pooling.

I think this entire file should be removed (or at least, the SED parts of it). I've spoken with eng about open-sourcing SED and they told me that it should go in esp-research. So this seems like it's kind of an open-sourcing of it, but it's not completely right (and doing it right is going to take some time). In addition to the pooling discrepencies, there is missing postprocessing steps on the SED outputs. 

Rather than trying to get the correct SED implementation here, we should probably just do it in esp-research. I think having two open-sourced versions will be confusing, especially if they don't function the same as each other.

File: avex/training/train.py [hunk 3]
```diff
@@ -133,6 +133,15 @@ def __init__(
         self.second_stage_lr: float | None = getattr(config.training_params, "second_stage_lr", None)
         self.second_stage_warmup_steps: int | None = getattr(config.training_params, "second_stage_warmup_steps", None)
         self._second_stage_active = self.freeze_backbone_epochs == 0  # Already unfrozen
+        self._prototypical_phase2_active: bool = False
```
Comment (benjaminsshoffman) [inline avex/training/train.py]: I don't see anything about differentiating learning rates for prototypes vs. head weights, am I missing it somewhere? This was something they did in audioprotopnet (lr for prototype was 0.05 and for head was 0.0005). I think this is quite important since I tried at one point training without this and it failed.

More broadly, have you run tests to see if prototype training is really working? (ie is performance high on benchmarks). Given there are some subtleties that might arise here, it seems essential to validate.

**Review by benjaminsshoffman — CHANGES_REQUESTED** [review]

File: avex/training/train.py (hunk 3 above)
Comment (nkundiushuti) [reply avex/training/train.py]: yes you are right, I can't trust the training part until I have one model running. there are two ways of training the model, the original paper and the perch2 way (gradient stop). the original way is done in two stages, the learning rate could be set separately for both. I am happy to leave out training for now...

File: examples/audioprotopnet_species_detection.py (hunk 2 above)
Comment (nkundiushuti) [reply examples/audioprotopnet_species_detection.py]: removed it

File: avex/data/ebird_taxonomy.py (hunk 1 above)
Comment (nkundiushuti) [reply avex/data/ebird_taxonomy.py]: I agree, I added an option to wire the corresponding ebird taxonomy to each model separately
